protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,376 @@
+// DEALINGS IN THE SOFTWARE.
+
+use crate::protocol::{AutoNatCodec, AutoNatProtocol, DialRequest, DialResponse, ResponseError};
+use libp2p_core::{
+    connection::{ConnectionId, ListenerId},
+    ConnectedPoint, Multiaddr, PeerId,
+};
+use libp2p_request_response::{
+    handler::RequestResponseHandlerEvent, ProtocolSupport, RequestResponse, RequestResponseConfig,
+    RequestResponseEvent, RequestResponseMessage, ResponseChannel,
+};
+use libp2p_swarm::{
+    AddressScore, DialPeerCondition, IntoProtocolsHandler, NetworkBehaviour,
+    NetworkBehaviourAction, PollParameters,
+};
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    iter,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+type FiniteAddrScore = u32;
+
+/// Config for the [`Behaviour`].
+pub struct Config {
+    // Timeout for requests.
+    timeout: Duration,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Config {
+            timeout: Duration::from_secs(5),
+        }
+    }
+}
+
+impl Config {
+    /// Set the timeout for dial-requests.
+    pub fn with_timeout(mut self, timeout: Duration) -> Self {
+        self.timeout = timeout;
+        self
+    }
+}
+
+/// Network Behaviour for AutoNAT.
+pub struct Behaviour {
+    // Inner protocol for sending requests and receiving the response.
+    inner: RequestResponse<AutoNatCodec>,
+    // Local listening addresses with a score indicating their reachability.
+    // The score increases each time a remote peer successfully dials this address.
+    addresses: HashMap<Multiaddr, FiniteAddrScore>,
+    // Ongoing inbound requests, where no response has been sent back to the remote yet.
+    ongoing_inbound: HashMap<PeerId, ResponseChannel<DialResponse>>,
+    // Ongoing outbound dial-requests, where no response has been received from the remote yet.
+    ongoing_outbound: HashSet<PeerId>,
+    // Recently connected peers to which we want to send a dial-request.
+    pending_requests: VecDeque<PeerId>,
+}
+
+impl Default for Behaviour {
+    fn default() -> Self {
+        Behaviour::new(Config::default())
+    }
+}
+
+impl Behaviour {
+    pub fn new(config: Config) -> Self {
+        let protocols = iter::once((AutoNatProtocol, ProtocolSupport::Full));
+        let mut cfg = RequestResponseConfig::default();
+        cfg.set_request_timeout(config.timeout);
+        let inner = RequestResponse::new(AutoNatCodec, protocols, cfg);
+        Self {
+            inner,
+            addresses: HashMap::default(),
+            ongoing_inbound: HashMap::default(),
+            ongoing_outbound: HashSet::default(),
+            pending_requests: VecDeque::default(),
+        }
+    }
+
+    /// Add a new address to the address list that is send to remote peers in a dial-request.
+    pub fn add_local_address(&mut self, address: Multiaddr) {
+        if self.addresses.get(&address).is_none() {
+            self.addresses.insert(address, 1);
+        }
+    }
```

Comment:
Okay, so I just noticed that `(Swarm)PollParameters::external_addresses`/ `registry::Addresses` already implements a majority of the logic that I was trying to achieve with my `local_address` map. I so far did not know about that, because the return type `AddressIntoIter` is not publicly exposed.  I assume that this is a mistake and that `AddressIntoIter` should be public; I can either change this as part of this PR, or do a separate one?  

But with this additional info, I would change quite a bit of my code, and I think I can remove the `addresses` HashMap completely, and just rely on the list provided in the `PollParameters::external_addresses` and send this list to the remote in a `DialRequest`.

> Just a thought: We could as well replace the current mechanism (PollParameters::external_addresses) through a push-style mechanism, e.g. introducing NetworkBehaviour::inject_external_address, e.g. triggered by a NetworkBehaviourAction::ReportObservedAddr emitted by another behaviour.

While I am in general more in favor of having such a push-style, I am not sure whether it makes sense in this case. Triggering `inject_external_address` would only be useful for the behaviour, if it independently maintains a map / list of its external addresses. I think in most cases this would just duplicate the logic that is already provided by `registry::Addresses`. Do you have specific use-cases in mind (maybe also from the other protocols) where this would make sense?
I think this is a general discussion, also kinda related to the discussion about the `AddressBook` in https://github.com/libp2p/rust-libp2p/discussions/2118: what should the Swarm implement in a central way, and what should be done independently by a Network Behaviour. I am a bit indecisive on this. 
But I think in general we need more documentation on the responsibilities/ concerns of the Swarm vs the ones of a Network Behaviour protocol.